### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries/PrimesInAP): proof of Dirichlet's Theorem

### PR DESCRIPTION
**Pull request by MichaelStollBayreuth**
---
This completes the proof of **Dirichlet's Theorem** on primes in residue classes.

See [here](https://leanprover.zulipchat.com/#narrow/channel/144837-PR-reviews/topic/Prerequisites.20for.20PNT.20and.20Dirichlet's.20Thm/near/484092915) on Zulip.

Note: I've added @loefflerd as an author in the `100.yaml` entry: without his great work on the analytic continuation of Dirichlet L-series, the proof would not had been possible.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean [hunk 1]
```diff
@@ -248,6 +288,193 @@ lemma LSeries_residueClass_eq (ha : IsUnit a) {s : ℂ} (hs : 1 < s.re) :
   simp only [Pi.smul_apply, residueClass_apply ha, smul_eq_mul, ← mul_assoc,
     mul_inv_cancel_of_invertible, one_mul, Finset.sum_apply, Pi.mul_apply]
 
+variable (a)
+
+open Classical in
+/-- The auxiliary function used, e.g., with the Wiener-Ikehara Theorem to prove
+Dirichlet's Theorem. On `re s > 1`, it agrees with the L-series of the von Mangoldt
+function restricted to the residue class `a : ZMod q` minus the principal part
+`(q.totient)⁻¹/(s-1)` of the pole at `s = 1`; see `DirichletsThm.auxFun_prop`. -/
```
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/PrimesInAP.lean:297]: It seems `auxFun_prop` is not part of this PR.

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean [hunk 2]
```diff
@@ -15,12 +15,52 @@ The goal of this file is to prove **Dirichlet's Theorem**: If `q` is a positive
...
+   (where `Λ` is the von Mangoldt function `ArithmeticFunction.vonMangoldt`; we have
+   `Λ (p^k) = log p` for prime powers and `Λ n = 0` otherwise.)
+2. Show that this function can be written as a linear combination of functions
+   of the form `χ * Λ` (pointwise product) with Dirichlet characters `χ` mod `q`.
+   See `ArithmeticFunction.vonMangoldt.residueClass_eq`.
+3. This implies that the L-series of `ArithmeticFunction.vonMangoldt.residueClass a`
+   agrees (on `re s > 1`) with the corresponding linear combination of negative logarithmic
+   derivatives of Dirichlet L-functions.
+   See `ArithmeticFunction.vonMangoldt.LSeries_residueClass_eq`.
+4. Define an auxiliary function `ArithmeticFunction.vonMangoldt.auxFun a` that is
```
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/PrimesInAP.lean:30]: It seems that `auxFun` plays quite an important role. I think it would be good if we can give it a more descriptive name. But I recognize that this is maybe very hard.

Do you have a suggestion? Or shall we ask on zulip for suggestions?

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean (hunk 1 above)
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/PrimesInAP.lean:297]: I changed `auxFun_prop` to `eqOn_auxFun` at some point, but forgot to update the docstring. Will fix.

Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/PrimesInAP.lean]: I found a few other places where a change in namespace was not reflected in the docstrings.

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean (hunk 2 above)
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/PrimesInAP.lean:30]: `Lfunction_residueClass_sub_inv_totient_div`? :smile: (It's a bit long...)

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean [hunk 3]
```diff
@@ -15,12 +15,55 @@ The goal of this file is to prove **Dirichlet's Theorem**: If `q` is a positive
...
+   this linear combination of negative logarithmic derivatives of L-functions minus
+   `(q.totient)⁻¹/(s-1)`, which cancels the pole at `s = 1`.
+   See `ArithmeticFuncion.vonMangolgt.eqOn_auxFun` for the statement that the auxiliary function
```
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/PrimesInAP.lean:33]:
```suggestion
   See `ArithmeticFuncion.vonMangoldt.eqOn_auxFun` for the statement that the auxiliary function
```

Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/PrimesInAP.lean]: There was also a "t" missing in "ArithmeticFuncion"...

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean (hunk 2 above)
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/PrimesInAP.lean]: Changed to `LfunctionResidueClassAux` as per @adomani's suggestion on  Zulip.

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean [hunk 4]
```diff
@@ -248,6 +291,201 @@ lemma LSeries_residueClass_eq (ha : IsUnit a) {s : ℂ} (hs : 1 < s.re) :
...
+-/
+
+section DirichletsTheorem
+
+namespace Nat
+
+open ArithmeticFunction vonMangoldt
+
+variable {q : ℕ} [NeZero q] {a : ZMod q}
+/-- **Dirichlet's Theorem** on primes in arithmetic progression: if `q` is a positive
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/LSeries/PrimesInAP.lean:471]:
```suggestion
variable {q : ℕ} [NeZero q] {a : ZMod q}

/-- **Dirichlet's Theorem** on primes in arithmetic progression: if `q` is a positive
```

**Review by riccardobrasca — APPROVED** [review]
Can you add a statement that uses `Infinite`, like in [Nat.infinite_setOf_prime_modEq_one](https://leanprover-community.github.io/mathlib4_docs/Mathlib/NumberTheory/PrimesCongruentOne.html#Nat.infinite_setOf_prime_modEq_one) (I am asking on zulip what we want to do with that file, now that the theorem is special case of Dirichlet).

File: Mathlib/NumberTheory/LSeries/PrimesInAP.lean [hunk 5]
```diff
@@ -248,6 +291,201 @@ lemma LSeries_residueClass_eq (ha : IsUnit a) {s : ℂ} (hs : 1 < s.re) :
...
+integer and `a : ZMod q` is a unit, then there are infintely many prime numbers `p`
+such that `(p : ZMod q) = a`. -/
+theorem setOf_prime_and_eq_mod_infinite (ha : IsUnit a) :
+    {p : ℕ | p.Prime ∧ (p : ZMod q) = a}.Infinite := by
```
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/PrimesInAP.lean:475]: @riccardobrasca is this the statement you are looking for? Should the name be permuted?

Comment (riccardobrasca) [reply Mathlib/NumberTheory/LSeries/PrimesInAP.lean:475]: Oh sorry, I somehow skip it in reading the PR! Forget about my comment (I don't have an opinion about the names)

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge
